### PR TITLE
feat: add Type ID implementation

### PR DESCRIPTION
**Pull request by XuJiandong**
* Add implementation of [Type ID](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0022-transaction-structure/0022-transaction-structure.md#type-id)
* Add new feature `type-id`
* Add test cases
* Update test dependencies

### CODE REVIEW

File: test/Cargo.toml [hunk 1]
```diff
@@ -13,3 +13,4 @@ serde_json = "1.0"
 ckb-mock-tx-types = "0.4.0"
 blake2b-rs = "0.1.5"
 faster-hex = "0.6"
+ckb-hash = "0.104.0"
```
Comment (mohanson) [inline test/Cargo.toml]: Why not use 0.118

File: src/type_id.rs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        return Err(SysError::TypeIDError);
+    }
+
+    // mint
+    if !is_cell_present(0, Source::GroupInput) {
+        let index = locate_index()?;
+        let input = load_input(0, Source::Input)?;
+        let mut blake2b = new_blake2b();
+        blake2b.update(input.as_slice());
+        blake2b.update(&index.to_le_bytes());
```
Comment (mohanson) [inline src/type_id.rs:82]: I suggest converting `index` from usize to u64 manually here, and then calling to_le_bytes

File: test/Cargo.toml (hunk 1 above)
Comment (XuJiandong) [reply test/Cargo.toml:16]: There will be another PR to update these dependencies. There are some minor changes to the project.

Comment (XuJiandong) [reply test/Cargo.toml]: Updated

File: src/error.rs [hunk 3]
```diff
@@ -20,7 +20,8 @@ pub enum SysError {
     MaxVmsSpawned,
     /// Max fds has been spawned. Its value is 9.
     MaxFdsCreated,
-
+    /// Type ID Error
+    TypeIDError,
```
Comment (mohanson) [inline src/error.rs:25]: Add `#[cfg(feature = "type-id")]`

**Review by mohanson — APPROVED** [review]
LGTM

**Review by xxuejie — APPROVED** [review]
In general this is fine but I'm confused here, it really seems to me that this PR mixes type ID logic with other non-type-id changes together. In the future it would be best if we can split non-related changes to multiple PRs.

**Comment by xxuejie** [comment]
It really seems to me that a different workflow is used here:

* Setup a common, separate branch elsewhere amongst a small team
* Everyone contributes to this separate branch together
* Periodically, the commits from this separate branch is merged over to upstream here

If this different workflow is used, it should be clearly noted in the PR title / description.

**Pull request merged**